File: 01/transp.js
// https://gist.github.com/femto113/1784503
function transpose(a)
{
  return a[0].map(function (_, c) { return a.map(function (r) { return r[c]; }); });
}

module.exports = {
  cifra: function(string, chave) {
    var matriz = [];
    var linha = -1;
    var i;

    for (i = 0; i < string.length; i++) {
      if (i % chave == 0) {
        matriz.push([]);
        linha++;
      }

      matriz[linha][i % chave] = string[i];
    }

    for (i = i % chave; i < chave; i++) {
      matriz[linha][i] = " ";
    }

    matriz = transpose(matriz);

    var str = "";

    for (var i = 0; i < matriz.length; i++) {
      for (var j = 0; j < matriz[0].length; j++) {
        str += matriz[i][j];
      }
    }

    return str;
  },
  decifra: function(string, chave) {
    var matriz = [];
    var linha = -1;
    var i;

    teto = string.length / chave;

    for (i = 0; i < string.length; i++) {
      if (i % teto == 0) {
        matriz.push([]);
        linha++;
      }

      matriz[linha][i % teto] = string[i];
    }

    var str = "";

    for (var j = 0; j < matriz[0].length; j++) {
      for (var i = 0; i < matriz.length; i++) {
        str += matriz[i][j];
      }
    }

    return str.trim();
  },
  chave: function(chave, callback) {
    callback(parseInt(chave, 10));
  }
}

File: cifras/transp.py
class Transposicao:
  def cifra(self, string, chave):
    matriz = []
    linha = -1
    i = 0

    for c in string:
      if i % chave == 0:
        linha += 1
        matriz.append([])

      matriz[linha].append(c)
      i += 1

    i = i % chave
    for i in range(i, chave):
      matriz[linha].append(" ")

    matriz = zip(*matriz);

    str = ""

    for i in range(0, len(matriz)):
      for j in range(0, len(matriz[0])):
        str += matriz[i][j]

    return str
  def decifra(self, string, chave):
    matriz = []
    linha = -1
    i = 0

    teto = len(string) / chave

    for c in string:
      if i % teto == 0:
        matriz.append([])
        linha += 1

      matriz[linha].append(c)
      i += 1

    r = i % teto
    if r > 0:
      while r < teto:
        matriz[linha].append(" ")
        r += 1

    str = []

    for j in range(0, len(matriz[0])):
      for i in range(0, len(matriz)):
        str.append(matriz[i][j])

    return "".join(str)
  def chave(self, chave):
    return int(chave)
  def quebraEmClaro(self, claro, criptografado):
    i = 1
    while 1:
      codigo = self.decifra(criptografado, i)
      diferente = False
      for j in range(0, len(codigo) - 2):
        if ord(codigo[j]) != ord(claro[j]):
          diferente = True
          break
      if not diferente:
        return i
      i += 1
  def quebraEscuro(self, criptografado, dicionario):
    ocorrencias = 0
    chave = 1

    for i in range(1, 10):
      codigo = self.decifra(criptografado, i)
      palavras = set(codigo.split())
      oco = len(palavras & set(dicionario))

      if oco > ocorrencias:
        chave = i
        ocorrencias = oco

    return chave
